Parquet_integration (#122)
cytomining/cytominer-database: README.rst

“batch” of images, and then later merge these CSVs.

cytominer-database ingest reads these CSVs, checks for errors, then ingests
them into a database backend, including
SQLite, MySQL, PostgresSQL, and several other backends supported by odo.

.. code-block:: sh

	cytominer-database ingest source_directory sqlite:///backend.sqlite -c ingest_config.ini

will ingest the CSV files nested under source_directory into a SQLite backend

How to use the configuration file
=================================
The configuration file ingest_config.ini must be located in the source_directory
and can be modified to specify the ingestion.
There are three different sections.

The [filenames] section
-----------------------

.. code-block::

  [filenames]
  image   = image.csv      #or: Image.csv
  object  = object.csv     #or: Object.csv

Cytominer-Database is currently limited to the following measurement file kinds:
Cells.csv, Cytoplasm.csv, Nuclei.csv, Image.csv, Object.csv.
The [filenames] section in the configuration file saves the correct basename of existing measurement files
(this may be important in the case of inconsistent capitalization).

> # TODO: Are there any other reasons for the [filenames] section?

Reviewer: make sure to remove

Author: removed comment

The [database_engine] section
-----------------------------

.. code-block::

  [ingestion_engine]
  engine = Parquet      #or: SQLite

The [database_engine] section specifies the backend. Possible key-value pairs are:
**engine** = *SQLite* or **engine** = *Parquet*.

> [Potential ToDo: Delete Section and introduce a command flag]

Reviewer: make sure to remove. If you're concerned about the thoughts being lost, then move them to a github issue

Author: removed [Potential ToDo: ...]

The [schema] section
--------------------

.. code-block::

 [schema]
 reference_option = sample         #or: path/to/reference/folder relative to source_directory
 ref_fraction     = 1              #or: any decimal value in [0, 1]
 type_conversion  = int2float      #or: all2string

The [schema] section specifies how to manage incompatibilities in the table schema
of the files.
In that case, a Parquet file is fixed to a schema with which it was first opened,
i.e. by the first file which is written (the reference file). To append the data
of all .csv files of that file-kind, it is important to assure the reference file
satisfies certain incompatibility requirements, e.g. does not miss any columns
and all existing files can be automatically converted to the reference schema.

> This section is used only if the files are ingested to Parquet format and was

Reviewer: this is a critical piece of this documentation. Can you highlight it some way?

Author: This only applies for the non-default case where the user decides to build a folder with reference files, instead of determining the reference files automatically. I highlighted that case with a warning-box (further below).

Author: Unfortunately, github does not support .rst warning boxes or note boxes (see also: [markup issue](https://github.com/github/markup/issues/68). For now I chose the strong emphasis from the [GitHub Markdown Specs](https://github.github.com/gfm/#emphasis-and-strong-emphasis). LMK if there's something more appealing you have in mind.

Reviewer: LOOKS GREAT!!!

developed to handle the special cases in which tables that cannot be concatenated automatically.

There are two options for the key **reference_option**:

> The first option is to create a designated folder containing one .csv reference file for every kind of file ("Cytoplasm.csv", "Nuclei.csv", ...) and save the folder path in the config file as **reference_option** = *path/to/reference/folder*, where the path is relative to the source_directory from the ingest command. These reference files' schema will determine the schema of the Parquet file into which all .csv files of its' kind will be ingested, hence the reference files must carefully selected to be complete in the number of columns and possess correct column names and types.

Reviewer: ```suggestion
The first option is to create a designated folder containing one .csv reference file for every kind of file ("Cytoplasm.csv", "Nuclei.csv", ...) and save the folder path in the config file as **reference_option** = *path/to/reference/folder*, where the path is relative to the source_directory from the ingest command.
These reference files' schema will determine the schema of the Parquet file into which all .csv files of its kind will be ingested.
Therefore, the reference files must be carefully selected to be complete in the number of columns and possess correct column names and types.
```

Note, this is a github "suggestion"

Reviewer: this means you can directly commit the change if you agree.

Alternatively, the reference files can be found automatically from a sampled subset of all existing files.
This is the case if **reference_option** = *sample* is set.
A subset of all files is sampled uniformly at random and the first table with
the maximum number of columns among all sampled .csv files is chosen as the reference table.
If this case, an additional key **ref_fraction** can be set, which specifies the fraction of files

> sampled among all files. The default value is **ref_fraction** = *1* , for which

Reviewer: Should this value be automatically set so high? What if I have a ton of images? Is it super quick to process and therefore ok to be set to 1 by default?

Author: For the release, some default value well below 1 would be better, if we can assume that almost all files do not have any missing columns and there are many files to ingest. The previous default value was set 1 because the number of testing files is small and hence even one or two corrupt .csv's could potentially dominate the outcome, if the sampling fraction is small. I double-checked all testing files and set the new default value to = 0.3 (tests still pass).

cytominer-database helps you organize these data into a single database backend, such as SQLite.

Why cytominer-database?
=======================
While tools like CellProfiler can store measurements directly in databases, it is usually infeasible to create a
centralized database in which to store these measurements. A more scalable approach is to create a set of CSVs per
“batch” of images, and then later merge these CSVs.

cytominer-database ingest reads these CSVs, checks for errors, then ingests
them into a database backend. The default backend is `SQLite`.

.. code-block:: sh

	cytominer-database ingest source_directory sqlite:///backend.sqlite -c ingest_config.ini

> will ingest the CSV files nested under source_directory into a `SQLite` backend

Reviewer: this is beautiful documentation

them into a database backend, including
SQLite, MySQL, PostgresSQL, and several other backends supported by odo.

.. code-block:: sh

	cytominer-database ingest source_directory sqlite:///backend.sqlite -c ingest_config.ini

will ingest the CSV files nested under source_directory into a SQLite backend

How to use the configuration file
=================================
The configuration file ingest_config.ini must be located in the source_directory
and can be modified to specify the ingestion.
There are three different sections.

The [filenames] section
-----------------------

.. code-block::

  [filenames]
  image   = image.csv      #or: Image.csv
  object  = object.csv     #or: Object.csv

Cytominer-Database is currently limited to the following measurement file kinds:
Cells.csv, Cytoplasm.csv, Nuclei.csv, Image.csv, Object.csv.
The [filenames] section in the configuration file saves the correct basename of existing measurement files
(this may be important in the case of inconsistent capitalization).

The [database_engine] section
-----------------------------

.. code-block::

  [ingestion_engine]
  engine = Parquet      #or: SQLite

> The [database_engine] section specifies the backend. Possible key-value pairs are:

Reviewer: ```suggestion
The [database_engine] section specifies the backend.
Possible key-value pairs are:
```

Quick fix for [one-sentence-per-line text philosophy](https://news.ycombinator.com/item?id=4642395). It makes text editing much easier in regards to github diffs. (and is a quick fix!)

cytominer-database helps you organize these data into a single database backend, such as SQLite.

Why cytominer-database?
=======================
While tools like CellProfiler can store measurements directly in databases, it is usually infeasible to create a
centralized database in which to store these measurements. A more scalable approach is to create a set of CSVs per
“batch” of images, and then later merge these CSVs.

cytominer-database ingest reads these CSVs, checks for errors, then ingests
them into a database backend, including

> SQLite, MySQL, PostgresSQL, and several other backends supported by odo.

Reviewer: ```suggestion
cytominer-database ingest reads these CSVs, checks for errors, then ingests them into a database backend.
cytominer-database supports SQLite and Parquet backends.
```

Wooooo! Updating lines 26 and 27 to be one-sentence-per-line. And, we no longer use `odo`

.. code-block:: sh

	cytominer-database ingest source_directory sqlite:///backend.sqlite -c ingest_config.ini

will ingest the CSV files nested under source_directory into a SQLite backend

How to use the configuration file
=================================
The configuration file ingest_config.ini must be located in the source_directory

> and can be modified to specify the ingestion.

Reviewer: ```suggestion
The configuration file ingest_config.ini must be located in the source_directory and can be modified to specify the ingestion.
```

Reviewer: [one sentence per line rule](https://news.ycombinator.com/item?id=4642395)

There are three different sections.

The [filenames] section
-----------------------

.. code-block::

  [filenames]
  image   = image.csv      #or: Image.csv
  object  = object.csv     #or: Object.csv

> Cytominer-Database is currently limited to the following measurement file kinds:

Reviewer: ```suggestion
cytominer-database is currently limited to the following measurement file kinds:
```

Cells.csv, Cytoplasm.csv, Nuclei.csv, Image.csv, Object.csv.
The [filenames] section in the configuration file saves the correct basename of existing measurement files

> (this may be important in the case of inconsistent capitalization).

Reviewer: ```suggestion
The [filenames] section in the configuration file saves the correct basename of existing measurement files.
This may be important in the case of inconsistent capitalization.
```

The [database_engine] section
-----------------------------

.. code-block::

  [ingestion_engine]
  engine = Parquet      #or: SQLite

The [database_engine] section specifies the backend.
Possible key-value pairs are:
**engine** = *SQLite* or **engine** = *Parquet*.

The [schema] section
--------------------

.. code-block::

 [schema]
 reference_option = sample         #or: path/to/reference/folder relative to source_directory
 ref_fraction     = 1              #or: any decimal value in [0, 1]
 type_conversion  = int2float      #or: all2string

The [schema] section specifies how to manage incompatibilities in the table schema
of the files.
In that case, a Parquet file is fixed to a schema with which it was first opened,
i.e. by the first file which is written (the reference file). To append the data
of all .csv files of that file-kind, it is important to assure the reference file
satisfies certain incompatibility requirements, e.g. does not miss any columns
and all existing files can be automatically converted to the reference schema.
This section is used only if the files are ingested to Parquet format and was

> developed to handle the special cases in which tables that cannot be concatenated automatically.

Reviewer: ```suggestion
The [schema] section specifies how to manage incompatibilities in the table schema of the files.
In that case, a Parquet file is fixed to a schema with which it was first opened, i.e. by the first file which is written (the reference file).
To append the data of all .csv files of that file-kind, it is important to assure the reference file satisfies certain incompatibility requirements.
For example, make sure the reference file does not miss any columns and all existing files can be automatically converted to the reference schema.
Note: This section is used only if the files are ingested to Parquet format and was developed to handle the special cases in which tables that cannot be concatenated automatically.
```

More one-sentence-per-line

We do this so that future changes to this PR can be handled easily without having to massively fix overflowing lines

There are two options for the key **reference_option**:
The first option is to create a designated folder containing one .csv reference file for every kind of file ("Cytoplasm.csv", "Nuclei.csv", ...)
and save the folder path in the config file as **reference_option** = *path/to/reference/folder*,
where the path is relative to the source_directory from the ingest command.

> These reference files' schema will determine the schema of the Parquet file into which all .csv files of its kind will be ingested.

Reviewer: ```suggestion
There are two options for the key **reference_option**:

The first option is to create a designated folder containing one .csv reference file for every kind of file ("Cytoplasm.csv", "Nuclei.csv", ...) and save the folder path in the config file as **reference_option** = *path/to/reference/folder*, where the path is relative to the source_directory from the ingest command.
These reference files' schema will determine the schema of the Parquet file into which all .csv files of its kind will be ingested.
```

**This option relies on manual selection, hence the chosen reference files must be checked explicitly: Make sure the .csv files are complete in the number of columns and contain no NaN values.**

Alternatively, the reference files can be found automatically from a sampled subset of all existing files.
This is the case if **reference_option** = *sample* is set.
A subset of all files is sampled uniformly at random and the first table with
the maximum number of columns among all sampled .csv files is chosen as the reference table.
If this case, an additional key **ref_fraction** can be set, which specifies the fraction of files
sampled among all files. The default value is **ref_fraction** = *1* , for which

> all tables are compared in width. This key is only used if "reference_option=sample".

Reviewer: ```suggestion
A subset of all files is sampled uniformly at random and the first table with the maximum number of columns among all sampled .csv files is chosen as the reference table.
If this case, an additional key **ref_fraction** can be set, which specifies the fraction of files sampled among all files.
The default value is **ref_fraction** = *1* , for which all tables are compared in width.
This key is only used if "reference_option=sample".
```

Lastly, the key **type_conversion** determines how the schema types are handled in the case of disagreement.
The default value is *int2float*, for which all integer columns are converted to floats.
This has been proven helpful for trivial columns (0-valued column), which may be of "int" type

> and cannot be written into the same table as non-trivial files with non-zero float values.

Reviewer: ```suggestion
This has been proven helpful for trivial columns (0-valued column), which may be of "int" type and cannot be written into the same table as non-trivial files with non-zero float values.
```
